Add protecting api blog (#2865)
FusionAuth/fusionauth-site: astro/src/content/blog/protecting-your-api.mdx

Having no authentication simplifies security and performance. It is a good option when no data modification needs to happen, or when there are read-only and write channels, as above.

However, such an approach is the exception rather than the rule. 

Far more often, especially when APIs allow clients to update data or offer functionality tied to the user's identity, you must authenticate and authorize the API request.

## Authentication And Authorization

Typically after you authenticate a request and know who the caller is, you authorize it. This lets the API server know what access the caller has or if access should be denied.

<Aside type="note" nodark="true">
Authentication is "who are you" and authorization is "what can you do". While they can be treated separately (car keys authorize without authenticating and a business card authenticates without authorizing), they are often considered together.
</Aside>

> Authorization can be done in a number of ways, including role-based access control (RBAC), attribute-based access control (ABAC), and relationship-based access control (ReBAC). Authorization decisions are typically embedded in business logic or extracted to a standalone policy decision point (PDP). 

Author: Agreed, but our primary solution for ABAC and ReBAC is "use an authorization server" like Cerbos or Permit.

With OAuth centralized API key verification, you can't rely on the central authority for everything. You must check the data it returns for correctness as well.

## First Party vs Third Party APIs

Another API security consideration is whether you are protecting a first-party or third-party API. In the example above, the todos client, the API server and the central authority are all owned by Pied Piper.

But what if Pied Piper builds a platform to foster an ecosystem of todos based add-ons. If this takes off, there are additional services which will make users' task management experience more enjoyable. Some examples of third-party todo based enhancements include:

* an LLM based service to summarize todos
* a reminder service to read todo due dates and text or email a user about upcoming tasks
* an analytics service to give a user monthly summaries of their accomplishments
* custom clients to add todos using an Apple watch interface

While all these services could be built by Pied Piper, if they open up a platform and provide the right incentives, they can harness the creativity and energy of more developers than they could ever hope to hire.

> With such a platform in mind, Pied Piper needs to authenticate not just the API requester, but also control access in a more sophisticated way, based on what each service needs. For example, the Apple watch client needs to be able to read and write todos, but the analytics service only needs to be able to read them.

Reviewer: ```suggestion
With such a platform in mind, Pied Piper needs to authenticate not just the API requester, but also control access in a more sophisticated way, based on what each service needs. For example, the Apple Watch client needs to be able to read and write todos, but the analytics service only needs to be able to read them.
```

For instance, if the analytics tool asks for write access to the todos, the request should be clear enough that any normal user will prohibit it. 

The central authority should also verify requested scopes. Since the scope request is under control of the third-party client, a malicious or buggy client could request disallowed values. If the central authority receives an initial scope request that was not granted to the third-party, the request is invalid and should not be processed.

Finally, if you are using scopes in a decentralized API key verification system, they should be validated by the API server to check that the scope in the API token matches the requested operation.

### Designing Scopes

Scope design is hard when first designing APIs, but even more painful to retrofit. Since third-party clients are building on top of the APIs you provide, updating or changing scopes will be a long, drawn out process.

> Spend some time up front thinking about what scopes make sense for your API. You want them to:

Reviewer: ```suggestion
Spend some time upfront thinking about what scopes make sense for your API. You want them to:
```

* be broad enough to cover a cohesive chunk of functionality
* be narrow enough to be useful to control access (avoid `superuser` scopes)
* be understandable to users
* have no overlap
* avoid being overbroad

Also, you don't need to have scopes for every API endpoint. For example, there may be certain admin APIs which would only be usable by first-party applications. In that case, don't write scope definitions for them. This avoids the possibility they'd ever be mistakenly granted to or requested by third-party applications. 

One way to design your scopes is to break your APIs into discrete functionality and offer read and write permissions to each area. For the Pied Piper Todo Management System, scopes might be:

* `todo-read`: the permission to read and list todos
* `todo-write`: the permission to create and update todos

As more API functionality is built, different read and write scopes can be added. It's also easier to add new scopes than to remove them, so err on the side of minimalism. This is more of an art than a science.

## Wrapping Up

When you are exposing an API, you need to protect it.

> You should think about whether to use a centralized API key verification system or decentralized verification. The centralized option requires constant communication between the consumers of API keys, the systems serving the API, and the central authority issuing API keys, all to ensure validity of the keys. 

Reviewer: ```suggestion
You should think about whether to use a centralized API key verification system or decentralized verification. The centralized option requires constant communication between the consumers of API keys, the systems serving the API, and the central authority issuing API keys, all to ensure the validity of the keys. 
```
